### DISCUSSION TITLE
FR: add text/markdown template to fileClass for full templating functionality

### DISCUSSION THREAD

**Discussion by Rhydderch** [mdelobelle/metadatamenu#590] [Ideas]
As I use metadataMenu as a way to manage templates, I was wondering whether it would be possible to also include markdown into the fileClass.

## Current set up and workflow
For now, metadataMenu only manages the properties of a file and gives the option to update properties/add missing properties. This is great to keep consistent properties across the same type of note. 

However, for the content of the note, I must rely on a markdown template and on the template plugin of obsidian.

So I have currently two templates for each file type

- **a fileClass**: templates properties
- **a markdown file**: contains a template for the body of the note and perhaps some scripts for the templater plugin

## Ideal workflow ✨ (it would be really amazing)
Allow people to add markdown to the fileClass and turn metadataMenu in **the** templating solution.

v1: allow users to add markdown that will be added once the fileClass is assigned to a file or on note creation (with no possibility of updates of the markdown; I guess this could be straightforward to implement?)

v2: use headers as structuring blocks and let users update the structure of notes based on the markdown template (e.g. add missing headers, but do not remove old headers)

<details><summary><b>Example image of what it would look like</b></summary>
<p>
<img width="610" alt="Screenshot 2024-01-16 at 11 07 21" src="https://github.com/mdelobelle/metadatamenu/assets/3197629/303d9640-d009-4232-8bee-fcaf2e7c5da0">

</p>

</details> 

What do you think? :)

**Comment by mdelobelle** [comment]
Hi I get your idea. and this is also a common situation that I've came across several times.
Since it's functionally speaking somewhat different from the purpose of metadata menu, I keep it open but not as a top priority. I'm currently focused on multi-files property editing (progressing slowly 😅) and then consider this idea

**Comment by mAAdhaTTah** [comment]
I would love this as well. One thing to flag in my usage of QuickAdd is file naming. I have a "person" type which I've configured QuickAdd to ask for First & Last name as separate fields, which get combined into the filename. While I would love to have the templates & fileClass feature combined into one plugin, there are definitely a lot of corner-y cases & features to resolve/include here.

**Comment by mdelobelle** [comment]
Hi, in order to keep things as simple as possible, I won't include this feature in the plugin. This really is the realm of templater and both plugin are running quite well together. In combination with Quickadd, you could definitely automate your workflow. 
There's a great demo of it on youtube by Dany Hatcher https://youtu.be/EBcZScQcDKg?si=F3UB3kavt-l7xE3c&t=1481

**Comment by Jawabiscuit** [comment]
Hey @Rhydderch, I support what you're asking for to a certain extent. I've expressed my thoughts on #543 about how MDM could fill the gap a little more while being cognizant of boundaries between MDM and plugins like TP. I suggest an API in that discussion to show the field modal and provide the consumer, e.g. TP, with the field formatted so that it's easily inserted into a new file's content wherever you choose. That's just how I envision it but it could certainly be taken in a different direction.

I have a working example of how I've glued up MDM with TP in the form of a sandbox vault that's currently private that I'll gladly document and make public if there's a demand here. In there, I've written modules, functionality provided by [obsidian-modules](https://github.com/polyipseity/obsidian-modules), to be an intermediary that rely on the settings stored in fileClass frontmatter in lieu of MDM providing an API. Having done that, I can confidently say that if MDM were to go in this direction, it will definitely spawn more issues and FRs for fields but nothing outlandish as far as I can tell and it would become even more exciting and powerful than it currently is.

The solution provided by modules is one that could break since it falls outside the perview of this plugin, but it provides a bridge for the time being. The power in what it provides is you're able to have a single "metatemplate", to borrow the term, that is activated on new file creation by TP and takes in a type as an argument, as opposed to having unique templates for each fileClass IMO making management unsustainable. Field values are prompted for by TP and content is filled in based on the type that was selected, which is what I think you are after @Rhydderch, kind of, but it only gets activated at creation time, not for notes already created. That template is 18 lines long though and provides a reasonable amount of flexibility over placement of content into the header, body and footer sections of the note.

Just a note in favor of some sort of templating: I think MDM could provide template fields since it seems, and I agree, fileClass templates are out of scope. This would make it possible for fields to hold inline DQL (`=this.property`), JS (`$= moment().format("YYYY-MM-DD")`) or DV views (`$= dv.view("myview")`). I personally would love templates for DV views e.g. for my breadcrumb navigation view: `$= dv.view("navigation", {file: "${"title"}"})` which gets formatted with the file title at creation time.

**Reply by ChristinWhite** [reply]
Interesting, I'd been playing with the idea of parsing Metadata Menu fileClass files and creating templates from them, probably with [Modal Form](https://github.com/danielo515/obsidian-modal-form) being somewhat of an intermediary. Modules might make things easier.

I think a multi-plugin approach may work really well, at least if you don't mind hacking things together.
